api/server/spanconverter.go

```diff
@@ -0,0 +1,129 @@
+package server
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+	"unicode"
+
+	"go.opencensus.io/stats"
+	view "go.opencensus.io/stats/view"
+	"go.opencensus.io/trace"
+)
+
+// SpanConverter registers as a opencensus Trace Exporter,
+// but it converts all the Spans in Views and registers them as such
+// A View exporter will then export them as normal.
+type SpanConverter struct {
+	opts     Options
+	measures map[string]*stats.Float64Measure
+	viewsMu  sync.Mutex
+	e        view.Exporter
+}
+
+// Options contains options for configuring the exporter.
+type Options struct {
+	Namespace string
+}
+
+func NewSpanConverter(o Options) (*SpanConverter, error) {
+	c := &SpanConverter{
+		opts:     o,
+		measures: make(map[string]*stats.Float64Measure),
+	}
+	return c, nil
+}
+
+var maxViews = 100
+
+// Spans are rejected if there are already maxViews (100) or they are
+// prefixed with '/', gin as been observed creating Span id specific
+// named Spans.
+func (c *SpanConverter) rejectSpan(sd *trace.SpanData) bool {
+	return len(c.measures) > maxViews || urlName(sd)
+}
+
+// ExportSpan creates a Measure and View once per Span.Name, registering
+// the View with the opencensus register. The length of time reported
+// by the span is then recorded using the measure.
+func (c *SpanConverter) ExportSpan(sd *trace.SpanData) {
+	if c.rejectSpan(sd) {
+		return
+	}
+	m := c.getMeasure(sd)
+
+	spanTimeNanos := sd.EndTime.Sub(sd.StartTime)
+	spanTimeMillis := float64(int64(spanTimeNanos / time.Millisecond))
+
+	stats.Record(context.Background(), m.M(spanTimeMillis))
+}
+
+func (c *SpanConverter) getMeasure(span *trace.SpanData) *stats.Float64Measure {
+	sig := sanitize(span.Name)
+	c.viewsMu.Lock()
+	m, ok := c.measures[sig]
+	c.viewsMu.Unlock()
+
+	if !ok {
```

Comment:
concurrent access without lock. Outside of locks this could register/insert more than once.

Reply:
Yeah it looked wrong to me to, when I lifted it from the main prom exporter. 

In actual fact there is no risk here, only that a view gets created twice, but OC will take care of that with it's deduplication logic.